crates/core/src/db/datastore/locking_tx_datastore/datastore.rs

```diff
@@ -626,6 +561,135 @@ impl MutProgrammable for Locking {
     }
 }
 
+#[derive(Debug, Error)]
+pub enum ReplayError {
+    #[error(transparent)]
+    Decode(#[from] bsatn::DecodeError),
+    #[error(transparent)]
+    Db(#[from] DBError),
+    #[error(transparent)]
+    Any(#[from] anyhow::Error),
+}
+
+/// A [`spacetimedb_commitlog::Decoder`] suitable for replaying a transaction
+/// history into the database state.
+pub struct Replay {
+    database_address: Address,
+    committed_state: Arc<RwLock<CommittedState>>,
+}
+
+impl spacetimedb_commitlog::Decoder for Replay {
+    type Record = spacetimedb_commitlog::payload::Txdata<ProductValue>;
+    type Error = spacetimedb_commitlog::payload::txdata::DecoderError<ReplayError>;
+
+    fn decode_record<'a, R: BufReader<'a>>(
+        &self,
+        version: u8,
+        tx_offset: u64,
+        reader: &mut R,
+    ) -> std::result::Result<Self::Record, Self::Error> {
+        let mut committed_state = self.committed_state.write_arc();
+        let mut visitor = ReplayVisitor {
+            database_address: &self.database_address,
+            committed_state: &mut committed_state,
+        };
+        spacetimedb_commitlog::payload::txdata::decode_record_fn(&mut visitor, version, tx_offset, reader)
+    }
+}
+
+struct ReplayVisitor<'a> {
+    database_address: &'a Address,
+    committed_state: &'a mut CommittedState,
+}
+
+impl spacetimedb_commitlog::payload::txdata::Visitor for ReplayVisitor<'_> {
+    type Error = ReplayError;
+    type Row = ProductValue;
+
+    fn visit_insert<'a, R: BufReader<'a>>(
+        &mut self,
+        table_id: TableId,
+        reader: &mut R,
+    ) -> std::result::Result<Self::Row, Self::Error> {
+        let schema = self
+            .committed_state
+            .schema_for_table(&ExecutionContext::default(), table_id)?
+            .into_owned();
+        let row = ProductValue::decode(schema.get_row_type(), reader)?;
```

Comment:
I implemented the encoding of the mutations according to [this](https://github.com/clockworklabs/SpacetimeDBPrivate/pull/718). It’s possible that we could just convert all rows per table to bflatn at once, instead of each row individually. But it’s not clear to me how exactly that would work.